File: python/extract-tiktok-streamlink.py
#!/usr/bin/env python3
# extract-tiktok-streamlink.sh — portiert nach python
# Quelle: shell, OpenClaw@gateway2:skills/tiktok-live-mon/scripts/extraction-methods/extract-tiktok-streamlink.sh
# Erzeugt: 2026-08-09 durch ABSTRACTIONS_MANAGER.py

import json
import os
import subprocess
import sys
from datetime import datetime

def emit_json(*args):
    keys = ("success", "method", "username", "url", "quality", "author", "title", "error", "timestamp", "status")
    values = args
    payload = {key: value for key, value in zip(keys, values) if value != ""}
    if "success" in payload:
        payload["success"] = payload["success"].lower() == "true"
    print(json.dumps(payload, ensure_ascii=False))

def validate_username(username):
    import re
    if not re.match(r'^[A-Za-z0-9._]{1,24}$', username):
        print("Invalid TikTok username", file=sys.stderr)
        sys.exit(64)

def validate_quality(quality):
    valid_qualities = ("best", "worst", "original", "1080p60", "720p60", "720p", "540p", "360p", "auto")
    if quality not in valid_qualities:
        print("Invalid stream quality", file=sys.stderr)
        sys.exit(64)

def get_load_per_cpu():
    load_avg = os.getloadavg()[0]
    cpu_count = max(1, os.cpu_count() or 1)
    return load_avg / cpu_count

def check_overload(max_load_per_cpu):
    load_per_cpu = get_load_per_cpu()
    if load_per_cpu > max_load_per_cpu:
        return True
    return False

def is_streamlink_installed():
    return subprocess.run(["which", "streamlink"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL).returncode == 0

def get_selector(quality):
    selector_map = {
        "original": "origin,uhd_60,hd_60,hd,sd,ld,best,worst",
        "auto": "best,origin,uhd_60,hd_60,hd,sd,ld,worst",
        "1080p60": "uhd_60,hd_60,hd,sd,ld,worst",
        "720p60": "hd_60,hd,sd,ld,worst",
        "720p": "hd,sd,ld,worst",
        "540p": "sd,ld,worst",
        "360p": "ld,worst"
    }
    return selector_map.get(quality, quality)

def run_streamlink_json(live_url, selector):
    try:
        result = subprocess.run(
            ["streamlink", "--json", live_url, selector],
            capture_output=True,
            text=True
        )
        if result.returncode == 0 and result.stdout:
            return json.loads(result.stdout)
        else:
            return None
    except Exception:
        return None

def run_streamlink_url(live_url, selector):
    try:
        result = subprocess.run(
            ["streamlink", "--stream-url", live_url, selector],
            capture_output=True,
            text=True
        )
        if result.returncode == 0 and result.stdout.strip():
            return result.stdout.strip()
        else:
            return None
    except Exception:
        return None

def parse_streamlink_json(data):
    url = data.get("url", "")
    streams = data.get("streams", {})
    if not url and isinstance(streams, dict):
        for key in ["best", "worst"] + list(streams.keys()):
            value = streams.get(key)
            if isinstance(value, dict) and value.get("url"):
                url = value["url"]
                break
    metadata = data.get("metadata", {})
    return {
        "url": url,
        "author": metadata.get("author", ""),
        "title": metadata.get("title", "")
    }

def main():
    if len(sys.argv) < 2:
        print("Usage: script.py <username> [quality] [--json]", file=sys.stderr)
        sys.exit(64)
    
    username = sys.argv[1].lstrip('@')
    quality = sys.argv[2] if len(sys.argv) > 2 and not sys.argv[2].startswith('--') else 'best'
    json_flag = '--json' if '--json' in sys.argv else ''
    timestamp = datetime.utcnow().strftime('%Y-%m-%dT%H:%M:%SZ')
    
    validate_username(username)
    validate_quality(quality)
    
    max_load_per_cpu = float(os.environ.get('TIKTOK_MAX_LOAD_PER_CPU', '1.5'))
    if check_overload(max_load_per_cpu):
        emit_json("false", "streamlink", username, "", quality, "", "", "host overloaded", timestamp, "overloaded")
        sys.exit(75)
    
    if not is_streamlink_installed():
        emit_json("false", "streamlink", username, "", quality, "", "", "streamlink not installed", timestamp, "dependency_missing")
        sys.exit(2)
    
    live_url = f"https://www.tiktok.com/@{username}/live"
    selector = get_selector(quality)
    
    output_data = run_streamlink_json(live_url, selector)
    if output_data is None:
        url = run_streamlink_url(live_url, selector)
        if not url:
            emit_json("false", "streamlink", username, "", quality, "", "", "streamlink failed or no stream found", timestamp, "offline")
            sys.exit(1)
        if json_flag:
            emit_json("true", "streamlink", username, url, quality, "", "", "", timestamp, "live")
        else:
            print(url)
        sys.exit(0)
    
    try:
        parsed = parse_streamlink_json(output_data)
    except Exception:
        emit_json("false", "streamlink", username, "", quality, "", "", "invalid streamlink JSON", timestamp, "technical_error")
        sys.exit(2)
    
    url = parsed.get("url", "")
    author = parsed.get("author", "")
    title = parsed.get("title", "")
    
    if not url:
        emit_json("false", "streamlink", username, "", quality, author, title, "could not extract stream URL", timestamp, "offline")
        sys.exit(1)
    
    if json_flag:
        emit_json("true", "streamlink", username, url, quality, author, title, "", timestamp, "live")
    else:
        print(url)

if __name__ == "__main__":
    main()

File: javascript/extract-tiktok-streamlink.js
#!/usr/bin/env node
// extract-tiktok-streamlink.sh — portiert nach javascript
// Quelle: shell, OpenClaw@gateway2:skills/tiktok-live-mon/scripts/extraction-methods/extract-tiktok-streamlink.sh
// Erzeugt: 2026-08-09 durch ABSTRACTIONS_MANAGER.py

import { spawnSync, execSync } from 'child_process';
import os from 'os';

const emitJson = (success, method, username, url, quality, author, title, error, timestamp, status) => {
    const payload = {
        success: typeof success === 'string' ? success.toLowerCase() === 'true' : success,
        method,
        username,
        url: url || undefined,
        quality,
        author: author || undefined,
        title: title || undefined,
        error: error || undefined,
        timestamp,
        status: status || undefined
    };
    
    // Remove undefined values
    Object.keys(payload).forEach(key => {
        if (payload[key] === undefined) {
            delete payload[key];
        }
    });
    
    console.error(JSON.stringify(payload));
};

const args = process.argv.slice(2);
const USERNAME = args[0] ? args[0].replace(/^@/, '') : '';
const QUALITY = args[1] || 'best';
const JSON_FLAG = args[2] || '';
const TIMESTAMP = new Date().toISOString().replace(/\.\d+Z$/, 'Z');

if (!USERNAME.match(/^[A-Za-z0-9._]{1,24}$/)) {
    console.error('Invalid TikTok username');
    process.exit(64);
}

if (!QUALITY.match(/^(best|worst|original|1080p60|720p60|720p|540p|360p|auto)$/)) {
    console.error('Invalid stream quality');
    process.exit(64);
}

const getLoadPerCpu = () => {
    const cpus = os.cpus().length || 1;
    const load = os.loadavg()[0];
    return load / cpus;
};

const LOAD_PER_CPU = getLoadPerCpu();
const MAX_LOAD = process.env.TIKTOK_MAX_LOAD_PER_CPU ? parseFloat(process.env.TIKTOK_MAX_LOAD_PER_CPU) : 1.5;

if (LOAD_PER_CPU > MAX_LOAD) {
    emitJson(false, 'streamlink', USERNAME, '', QUALITY, '', '', 'host overloaded', TIMESTAMP, 'overloaded');
    process.exit(75);
}

const streamlinkExists = () => {
    try {
        execSync('which streamlink', { stdio: 'ignore' });
        return true;
    } catch {
        return false;
    }
};

if (!streamlinkExists()) {
    emitJson(false, 'streamlink', USERNAME, '', QUALITY, '', '', 'streamlink not installed', TIMESTAMP, 'dependency_missing');
    process.exit(2);
}

const LIVE_URL = `https://www.tiktok.com/@${USERNAME}/live`;

const QUALITY_SELECTOR = {
    original: 'origin,uhd_60,hd_60,hd,sd,ld,best,worst',
    auto: 'best,origin,uhd_60,hd_60,hd,sd,ld,worst',
    '1080p60': 'uhd_60,hd_60,hd,sd,ld,worst',
    '720p60': 'hd_60,hd,sd,ld,worst',
    '720p': 'hd,sd,ld,worst',
    '540p': 'sd,ld,worst',
    '360p': 'ld,worst'
};

const SELECTOR = QUALITY_SELECTOR[QUALITY] || QUALITY;

const runStreamlink = (args) => {
    const result = spawnSync('streamlink', args, { encoding: 'utf8', stdio: ['pipe', 'pipe', 'pipe'] });
    return {
        stdout: result.stdout.trim(),
        stderr: result.stderr.trim(),
        status: result.status
    };
};

const jsonResult = runStreamlink(['--json', LIVE_URL, SELECTOR]);
let EXIT_CODE = jsonResult.status;
let OUTPUT = jsonResult.stdout;

if (EXIT_CODE !== 0 || !OUTPUT) {
    const urlResult = runStreamlink(['--stream-url', LIVE_URL, SELECTOR]);
    const URL = urlResult.stdout;
    
    if (urlResult.status !== 0 || !URL) {
        emitJson(false, 'streamlink', USERNAME, '', QUALITY, '', '', 'streamlink failed or no stream found', TIMESTAMP, 'offline');
        process.exit(1);
    }
    
    if (JSON_FLAG === '--json') {
        emitJson(true, 'streamlink', USERNAME, URL, QUALITY, '', '', '', TIMESTAMP, 'live');
    } else {
        console.log(URL);
    }
    process.exit(0);
}

let PARSED;
try {
    const data = JSON.parse(OUTPUT);
    let url = data.url || '';
    const streams = data.streams || {};
    
    if (!url && typeof streams === 'object') {
        const keys = ['best', 'worst', ...Object.keys(streams)];
        for (const key of keys) {
            const value = streams[key];
            if (typeof value === 'object' && value.url) {
                url = value.url;
                break;
            }
        }
    }
    
    const metadata = data.metadata || {};
    PARSED = {
        url: url,
        author: metadata.author || '',
        title: metadata.title || ''
    };
} catch (e) {
    emitJson(false, 'streamlink', USERNAME, '', QUALITY, '', '', 'invalid streamlink JSON', TIMESTAMP, 'technical_error');
    process.exit(2);
}

const { url: URL, author: AUTHOR, title: TITLE } = PARSED;

if (!URL) {
    emitJson(false, 'streamlink', USERNAME, '', QUALITY, AUTHOR, TITLE, 'could not extract stream URL', TIMESTAMP, 'offline');
    process.exit(1);
}

if (JSON_FLAG === '--json') {
    emitJson(true, 'streamlink', USERNAME, URL, QUALITY, AUTHOR, TITLE, '', TIMESTAMP, 'live');
} else {
    console.log(URL);
}
